### ISSUE TITLE
Rename useSetRecoilState() to useSetRecoilValue()

### ISSUE THREAD

**Issue by dharkness** [facebookexperimental/Recoil#270]
The name `useSetRecoilState` doesn't match the other primary hooks. In several places, the documentation refers to "the value of the state" which implies that the state is more than just the value. And while the docs also refer to the value as the state, the names of these hooks don't go together.

* `useRecoilState` - retrieve the value and the function to set the value
* `useRecoilValue` - retrieve the value
* `useSetRecoilState` - retrieve the function to set the value

The description of `useSetRecoilState` makes it more clear that it operates on the value rather than the full state.

> Returns a setter function for updating the value of writeable Recoil state.

Or does this hook allow you to do more than set the value?

**Comment by drarmstr** [comment]
The hook names consistently use the terms `RecoilState` and `RecoilValue` to indicate the types they accept as a parameter.  There are the following types:

* `RecoilState` - Writeable value
* `RecoilValueReadOnly` - Read-only value
* `RecoilValue` - Writeable or Read-only value

So, the following hooks require writeable type:

* `useRecoilState`
* `useSetRecoilState`
* `useResetRecoilState`
* `useRecoilStateLoadable`

While the follow can accept either read-only or writeable type:

* `useRecoilValue`
* `useRecoilValueLoadble`

**Comment by dharkness** [comment]
Thank you for the detailed explanation. As I'm reading through more, I can see the logic. Apologies upfront if this comment seems to ramble a bit. I anticipate repeated Stack Overflow questions like these and am trying to help avoid them.

* I see `useRecoilValue`, but where is `useSetRecoilValue`?
* Why doesn't `const [ fontSizeLabel ] = useRecoilState(fontSizeLabelState)` work?

> TLDR
> I suppose as long as there's a paragraph in the **Core Concepts** section that clearly differentiates between _state_ and _value_—and the docs stick to that usage—it should be fine.

It still feels weird to me that
```
type RecoilValue = RecoilValueReadOnly | RecoilState
```
When I think of React and Redux state, I think of the value as the current raw JavaScript value and the state as the definition of some value we want to track. `fontSizeState` is the state whereas `12` is its value. But with these Recoil types, state is a value that is not read-only, and value is a writable or read-only state. It all feels very self-referential. ;)

Renaming the types might clear things up.

* `RecoilWritable` - writable state
* `RecoilReadOnly` - read-only state
* `RecoilReadable` - any writable or read-only state

They would also remove the confusion between "value" as a type, as a core concept, and in the hook names. It's immediately clear why `isRecoilValue()` and `useRecoilValue()` accept the relaxed `RecoilReadable` and why `useSetRecoilState()` isn't `useSetRecoilValue()`.

I don't think you'd have to rename the hooks to match the type names, but it sure makes their usage much clearer to me.

* `useRecoilState` -> `useRecoilWritable`
* `useRecoilValue` -> `useRecoilReadable`
* `useSetRecoilState` -> `useSetRecoilWritable`
* `useResetRecoilState` -> `useResetRecoilWritable`

Or heck, make `Recoil` synonymous with `State` and keep `Value` to indicate read-only?

* `isRecoilValue` -> `isRecoil`
* `useRecoilState` -> `useRecoil`
* `useRecoilValue` -> `useRecoilValue` (unchanged)
* `useSetRecoilState` -> `useSetRecoil`
* `useResetRecoilState` -> `useResetRecoil`

I admit that might paint you into a corner when you need to introduce other Recoil concepts and hooks.

Cheers!

**Comment by drarmstr** [comment]
Yes, the type naming is a little interesting and we have many discussions about it...  ;) Particularly since the read/write concept is related but different than the atom/selector distinction.  Another naming constraint is that we wanted `useRecoilState()` to have a similar name extending `useState()`, as it has a similar signature and is the first thing to explain to people comparing with React state.

**Comment by dharkness** [comment]
For sure, I understand the desire to make parallel hooks for ease of transition. I had a lot more phrases like "I think that" or "This probably won't work" as I initially wrote my reply. :) I really like the new library as it solves a lot of the boilerplate inconvenience of Redux, though I do think that's often overblown since I can copy-paste `store.js` without modification into each project.

And I like that it allows the same sort of hook APIs I've created for Redux, e.g. `useLeadType()` that returns the value and setter and `useSetLeadType()` that returns only the setter to avoid re-renders when you don't need the current value.
